### ISSUE TITLE
Adding properties causes designer to crash

### ISSUE THREAD

**Issue by vytsci** [nocode-js/sequential-workflow-editor#49]
Hello, I am new here, tried to find a documented solution with no avail.

I have an issue every time I add a new property, old schemas break, and I need to add this property to all my old schemas.

Is it a bug and is there some workaround for this? Or does this problem have a proper solution similar to database migrations or smth?

Either way it should be documented and have an infrastructure for this.

**Comment by b4rtaz** [comment]
Hello @vytsci, I've just added a new article to nocode-js.com: https://nocode-js.com/docs/sequential-workflow-editor/updating-model I hope it will be helpfull.

**Comment by vytsci** [comment]
@b4rtaz thanks, that will do for now. Sad it throws error instead a controllable message so you can ask user to assign missing properties in root sidebar.

**Comment by vytsci** [comment]
@b4rtaz came up with a dynamic solution, maybe if you see any upcoming issues, you can help me resolve them?

```import {PropertyModel} from "sequential-workflow-editor-model"
import {definitionModel, MyDefinition} from "@gUtils/sequence-builder/model/definition-model"
import {DefinitionWalker} from "sequential-workflow-model"
import {cloneDeep} from "lodash";

export class DefinitionMigrator {
  public static migrate(definition: MyDefinition): MyDefinition {
    const migratedDefinition = cloneDeep(definition);

    const walker = new DefinitionWalker();

    walker.forEach(migratedDefinition, step => {
      const stepModel = definitionModel.steps[step.type];
      definitionModel.steps
      stepModel.properties.forEach((stepModelProperty: PropertyModel) => {
        if (step.properties[stepModelProperty.path.last()] === undefined) {
          step.properties[stepModelProperty.path.last()] = {
            modelId: stepModelProperty.value.subModels?.length
              ? stepModelProperty.value.subModels[0].id
              : stepModelProperty.value.id,
            value: null
          };
        }
      });
    });

    return migratedDefinition;
  }
}
```

**Comment by b4rtaz** [comment]
> thanks, that will do for now. Sad it throws error instead a controllable message so you can ask user to assign missing properties in root sidebar.

Yeah, basically, this problem is a bit more complex from my perspective. Let’s assume you have n different steps, and at some point, you need to add a new property. At first glance, this seems straightforward. I had the idea to introduce a utility in the SWE framework to handle this, like `.fallback(value)`, for example:

```ts
myStep.property('username')
   .value(createStringValueModel({}))
   .fallback('default value for upgraded steps to a new version of model');
```

This approach seems to solve the problem, but it introduces several deeper issues. The first issue I noticed is that the change is often more complicated. If you modify one property, there’s a high chance another property is also affected. What if these values are not independent? Or what if you want to combine the `username`, `password`, and `dbName` properties into a single `connectionString` property? In many cases, it’s impossible to set a default value because no universal default exists, the user has to define it. And there are many other use cases to consider. Sometimes just creating a new version of a step is only a solution. Sometimes not.

From this perspective, I haven’t been able to find a good solution that can be applied at the framework level. Some projects only require simple update logic, while more advanced projects need complex solutions, such as upgrading storage, versioning, and so on.

I also think that if a developer can easily set a fallback, in a large project, no one will understand what’s happening. Let’s consider this scenario: the first release includes the following code:

```ts
myStep.property('color')./*...*/.fallback('red');
```

The next release includes this code:

```ts
myStep.property('color')./*...*/.fallback('blue');
```

Now, if you have 1,000 definitions created by your users, you have no way of knowing which definition was updated by which version.

The worst part is that this problem grows with the number of steps and the size of your development team. That’s why I recommend choosing a solution tailored to your specific project.
